### ISSUE TITLE
Automatically tidy up system indices metadata other than mappings

### ISSUE THREAD

**Issue by williamrandolph** [elastic/elasticsearch#86302]
The have been a few issues recently where we've fixed a bug in system indices or system data streams in a way that prevents a problem on newly created indices. For example:

* https://github.com/elastic/elasticsearch/pull/85511#pullrequestreview-926838060
* https://github.com/elastic/elasticsearch/pull/85977

It is not always clear if these fixes will be propagated to existing system indices by the `SystemIndexManager` cluster state listener; looking at the code, it appears that we only really handle the case of mappings. We can fix these issues for 7.x indices on an 8.x cluster by running a system feature migration, but this can cause its own problems. For example:

* https://github.com/elastic/elasticsearch/pull/85792

In the design of system indices, we did a lot of work to make sure that system indices could be created correctly in mixed-version clusters. A system index descriptor can hold a reference to system index descriptor for a previous cluster version. If this were implemented, an 8.3 node that receives a request to create a system index in a cluster with a minimum node version of 8.1 might find that the system index descriptor returned by a plugin has a minimum version of 8.3, but that the `getDescriptorCompatibleWith` method can return a different system index descriptor appropriate for 8.1. Once the cluster is fully upgraded, the `SystemIndexManager` object can determine that mappings are out of date and upgrade them automatically.

But as far as I can tell, none of our system index descriptors actually hold these prior index descriptors, so I don't think we're really using this mechanism.

Is this something we can leverage to propagate bugfixes to non-mapping system index metadata, like whether the index is hidden or what its aliases are? Or is there a quicker approach we could take to fixing this?

**Comment by williamrandolph** [comment]
I poked at the code a little bit more. There's no standard mechanism for creating system indices. Some system indices, like `.tasks` and `.logstash`, are auto-created when the code uses a `Client` to index a document to the system index. Others, like the security system index, are created by plugin-specific index managers that listen for cluster state changes.

Both the normal creation and autocreation code paths try to find a system index descriptor that is compatible with the minimum node version in the cluster. They do this by searching a system index descriptors list of "prior descriptors." In practice, this never happens because we have not made prior index descriptors for any of the system indices.

Once the index is created, `SystemIndexManager` will update its mappings. Before doing this, it checks that the cluster is not a mixed-version cluster. This only runs on the master node.

Things we might want to clean up:
1. If a system index is not hidden, make it hidden. (is this actually a dynamic setting?)
2. If a system index's settings have changed, update them.
3. If a primary system index is not the write index for its alias, make it the write index.
4. Others? Nothing is coming to mind right now.

The difficulty with changing arbitrary settings is that we would have to distinguish between fixed and dynamic settings. We may simply want to require a reindexing, probably managed through the "system feature migration" API.

**Comment by williamrandolph** [comment]
This is embarrassing. We do have a class that does this: it's `SystemIndexMetadataUpgradeService`, and it was added way back in https://github.com/elastic/elasticsearch/pull/57936/. I worked on it myself when making system indices into hidden indices.

These classes need to be merged or at least linked together via Javadoc. I'll look at the issues linked above and see if I can find any specific bug; if so, I'll file a new issue.
